v3/src/components/axis/hooks/use-axis.ts

```diff
@@ -17,124 +17,239 @@ export interface IUseAxis {
   titleRef: MutableRefObject<SVGGElement | null>
   label?: string
   enableAnimation: MutableRefObject<boolean>
-  showGridLines: boolean
+  showScatterPlotGridLines: boolean
+  centerCategoryLabels: boolean
 }
 
 export const useAxis = ({
                           axisModel, axisElt, titleRef, label = "",
-                          showGridLines, enableAnimation
+                          showScatterPlotGridLines, centerCategoryLabels, enableAnimation
                         }: IUseAxis) => {
   const layout = useAxisLayoutContext(),
+    isNumeric = axisModel?.isNumeric,
     place = axisModel?.place ?? 'bottom',
     scale = layout.getAxisScale(place) as ScaleNumericBaseType,
+    ordinalScale = isNumeric || axisModel?.type === 'empty' ? null : scale as unknown as ScaleBand<string>,
+    bandWidth = ordinalScale?.bandwidth() ?? 0,
     axis = (place === 'bottom') ? axisBottom : axisLeft,
-    isNumeric = axisModel?.isNumeric,
     // By all rights, the following three lines should not be necessary to get installDomainSync to run when
     // GraphController:processV2Document installs a new axis model.
     // Todo: Revisit and figure out whether we can remove the workaround.
     previousAxisModel = useRef<IAxisModel>(),
     axisModelChanged = previousAxisModel.current !== axisModel,
     dataConfiguration = useDataConfigurationContext(),
     axisPlace = axisModel?.place ?? 'bottom',
-    attrRole = graphPlaceToAttrPlace(axisPlace),
-    [xMin, xMax] = scale?.range() || [0, 100],
-    halfRange = Math.abs(xMax - xMin) / 2,
+    attrRole = graphPlaceToAttrRole(axisPlace),
+    [rangeMin, rangeMax] = scale?.range() || [0, 100],
+    halfRange = (rangeMax !== undefined && rangeMin !== undefined) ? Math.abs(rangeMax - rangeMin) / 2 : 50,
```

Comment:
Unfortunately scale?.range() evaluates to an empty array in early stages of starting up and thus both rangeMax and rangeMin come out as undefined. I was able to slightly improve things with:
range = scale?.range() || [0, 100],
[rangeMin, rangeMax] = range.length === 2 ? range : [0, 100],
halfRange = Math.abs(rangeMax - rangeMin) / 2,

Reply:
Ah, then there's a bug in the TypeScript types for D3, because the existing types suggest that `rangeMin` and `rangeMax` are always numeric, i.e. can't be `undefined`.